src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,370 @@
+    // Assertion stack.
+    Collection<Term> extraAssertions = getAssertedTermsNotInCollection(combinedInterpols);
+
+    // Uses a separate Solver instance to leave the original solver-context unmodified
+    Solver interpolationSolver = new Solver();
+
+    setSolverOptions(seed, solverOptions, interpolationSolver);
+
+    Term interpolantA = buildConjunctionOfFormulasOverList(pFormAAsList, interpolationSolver);
+
+    if (!extraAssertions.isEmpty()) {
+      Term extraAssert = buildConjunctionOfFormulas(extraAssertions, interpolationSolver);
+      interpolantA = interpolationSolver.mkTerm(Kind.AND, extraAssert, interpolantA);
+    }
+    Term interpolantB = buildConjunctionOfFormulasOverList(pFormBAsList, interpolationSolver);
+
+    interpolationSolver.assertFormula(interpolantA);
+
+    Term interpolant =
+        interpolationSolver.getInterpolant(interpolationSolver.mkTerm(Kind.NOT, interpolantB));
+
+    interpolationSolver.deletePointer();
+
+    return interpolant;
+  }
+
+  /**
+   * Returns Asserted Formulas (from the formula Stack), not present in the given collection of
+   * Terms.
+   *
+   * @param pCombinedInterpols asserted Formulas to invert the selection
+   * @return asserted Formulas not in collTerms, but in the formula Stack
+   */
+  private Collection<Term> getAssertedTermsNotInCollection(
+      ImmutableList<Collection<Term>> pCombinedInterpols) {
+    ImmutableList<Term> retTerms =
+        pCombinedInterpols.stream()
+            .flatMap(Collection::stream)
+            .collect(ImmutableList.toImmutableList());
+
+    Set<Term> filteredAssertedTerms =
+        assertedFormulas.stream()
+            .flatMap(c -> c.stream())
+            .filter(n -> !retTerms.contains(n))
+            .collect(ImmutableSet.toImmutableSet());
+    return filteredAssertedTerms;
+  }
+
+  /**
+   * Turns a List of Collections of Formulas to a Single Conjunction of the Formulas e.g.:
+   * [[A,B],[C]] -> A/\B/\C.
+   *
+   * @param formula List of Collections of formulas
+   * @param usingSolver the CVC5 Solver Instance to use
+   * @return concatenated Formulas with AND as CVC5 Term
+   */
+  private Term buildConjunctionOfFormulasOverList(
+      ImmutableList<Collection<Term>> formula, Solver usingSolver) {
+    Collection<Term> eachCollectionOfTermsConcat =
+        formula.stream()
+            .map(n -> buildConjunctionOfFormulas(n, usingSolver))
+            .collect(ImmutableList.toImmutableList());
+    Term concatTerm = buildConjunctionOfFormulas(eachCollectionOfTermsConcat, usingSolver);
+
+    return concatTerm;
+  }
+
+  /**
+   * Turns a collection of Formulas to a Single Conjunction of the Formulas e.g.: [A,B,C] ->
+   * A/\B/\C.
+   *
+   * @param formulas collection of formulas
+   * @param usingSolver the CVC5 Solver Instance to use
+   * @return concatenated Formulas with AND as CVC5 Term
+   */
+  private Term buildConjunctionOfFormulas(Collection<Term> formulas, Solver usingSolver) {
```

Comment:
This method looks like a very expensive version of 
```
switch(formulas.size()) {
  case 0: return solver.mkBoolean(true);
  case 1: return Iterables.getOnlyElement(formulas);
  default: return solver.mkTerm(Kind.AND, formulas.toArray());
}
```

If possible, take a look into the implementation of CVC5BooleanFormulaManager:makeAnd(Collection).

Reply:
Good Point. Thanks.